Implementation of the mpio driver with selection I/O.

vchoi-hdfgroup opened the pull request:
This also addresses the review comments of the previous PR #3058 which will be closed.

vchoi-hdfgroup committed:
Implementation of the mpio driver with selection I/O.
This also addresses the review comments of the previous PR #3058 which will be closed.

github-actions[bot] committed:
Committing clang-format changes

vchoi-hdfgroup committed:
Add comments for the changes. Minor modifications to the mpio driver for selection I/O.

github-actions[bot] committed:
Committing clang-format changes

vchoi-hdfgroup committed:
Correct spelling error.

vchoi-hdfgroup committed:
Merge branch 'HDFGroup:develop' into new_mpio_develop

vchoi-hdfgroup committed:
Merge branch 'HDFGroup:develop' into new_mpio_develop

vchoi-hdfgroup committed:
Merge branch 'HDFGroup:develop' into new_mpio_develop

vchoi-hdfgroup committed:
Merge branch 'HDFGroup:develop' into new_mpio_develop

vchoi-hdfgroup committed:
Add tests for the three bugs found when setting H5D_SELECTION_IO_MODE_ON as
the default in the library.  Minor cleanups.

github-actions[bot] committed:
Committing clang-format changes

vchoi-hdfgroup committed:
Correct spelling error.

vchoi-hdfgroup committed:
Merge branch 'HDFGroup:develop' into new_mpio_develop

vchoi-hdfgroup committed:
Merge branch 'HDFGroup:develop' into new_mpio_develop

vchoi-hdfgroup committed:
Merge branch 'HDFGroup:develop' into new_mpio_develop

vchoi-hdfgroup committed:
Add tests for selection I/O which are derived from test/vfd.c and modified for parallel testing.

github-actions[bot] committed:
Committing clang-format changes

vchoi-hdfgroup committed:
Correct spelling errors.
Remove debugging printfs.

github-actions[bot] committed:
Committing clang-format changes

vchoi-hdfgroup committed:
Merge branch 'HDFGroup:develop' into new_mpio_develop

vchoi-hdfgroup committed:
Merge branch 'develop' into new_mpio_develop

github-actions[bot] committed:
Committing clang-format changes

fortnern commented on src/H5FDdevelop.h:
```diff
@@ -196,12 +196,12 @@ typedef struct H5FD_class_t {
                           size_t sizes[], void *bufs[]);
     herr_t (*write_vector)(H5FD_t *file, hid_t dxpl, uint32_t count, H5FD_mem_t types[], haddr_t addrs[],
                            size_t sizes[], const void *bufs[]);
-    herr_t (*read_selection)(H5FD_t *file, H5FD_mem_t type, hid_t dxpl_id, size_t count, hid_t mem_spaces[],
```
I understand we want to keep things consistent but if we're going to change the API to keep the count parameter the same type we should change it to size_t. I think we should just lave it alone for now though.

vchoi-hdfgroup replied:
So I will keep <count > as size_t for read_selection() and write_selection()? But <count> as uint32_t for write_vector and read_vector?

fortnern replied:
Yeah I don't think we should change it now, it's already part of the API

vchoi-hdfgroup replied:
Just want to double-check:
<count> as size_t will be used for all APIs and the related intermediate routines:
H5FDread/write_selection(), H5FDread/write_vector_from_selection(), H5FDread/write_from_selection()

vchoi-hdfgroup replied:
correction to previous comment: "count" as size_t

fortnern replied:
Don't change any APIs right now, thanks, just revert the type change to the H5FD_class_t struct

fortnern commented on src/H5FDmpio.c:
```diff
@@ -2716,6 +2733,924 @@ H5FD__mpio_write_vector(H5FD_t *_file, hid_t H5_ATTR_UNUSED dxpl_id, uint32_t co
+            }
+        }
+
+        if (H5FD_read_vector_from_selection(_file, type, count, mem_space_ids, file_space_ids, offsets,
```
I think this should actually be H5FD_read_from_selection(). Same for write (I know my earlier email said vector)

vchoi-hdfgroup replied:
Done

vchoi-hdfgroup committed:
Merge branch 'HDFGroup:develop' into new_mpio_develop

vchoi-hdfgroup committed:
Fix CI failures due ot ";" after macros.

github-actions[bot] committed:
Committing clang-format changes

vchoi-hdfgroup committed:
This changes the default selection I/O to on for MPIO.
This is a fix from Neil.

vchoi-hdfgroup committed:
Merge branch 'HDFGroup:develop' into new_mpio_develop

vchoi-hdfgroup committed:
Fixes for the two issues in PR review:
1) Revert the type for <count> as size_t in the H5FD_class_t struct
2) The independent xfer mode case in H5FD__mpio_read/write_selection() should be
calling H5FD_read/write_from_selection().
3) Fix the test in testpar/t_select_io_dset.c due to the change.

github-actions[bot] committed:
Committing clang-format changes

lrknox commented on src/H5FD.c:
```diff
@@ -1820,14 +1820,408 @@ H5FDwrite_selection(H5FD_t *file, H5FD_mem_t type, hid_t dxpl_id, uint32_t count
+ *
+ * Return:      Success:    SUCCEED
+ *                          All reads have completed successfully, and
+ *                          the results havce been into the supplied
```
```suggestion
 *                          the results have been written into the supplied
```

lrknox commented on src/H5FD.c:
```diff
@@ -1820,14 +1820,408 @@ H5FDwrite_selection(H5FD_t *file, H5FD_mem_t type, hid_t dxpl_id, uint32_t count
+ *
+ * Return:      Success:    SUCCEED
+ *                          All reads have completed successfully, and
+ *                          the results havce been into the supplied
```
```suggestion
 *                          the results havce been into the supplied
```

lrknox commented on src/H5FDint.c:
```diff
@@ -1999,6 +2033,217 @@ H5FD_write_selection_id(H5FD_t *file, H5FD_mem_t type, uint32_t count, hid_t mem
+ *              Failure:    FAIL
+ *                          One or more writes failed.
+ *
+ * Programmer:  NAF -- 5/19/21
```
Are we removing "Programmer" references?

lrknox commented on src/H5FDint.c:
```diff
@@ -2378,6 +2671,187 @@ H5FD_sort_vector_io_req(hbool_t *vector_was_sorted, uint32_t _count, H5FD_mem_t
+ *              of the selection I/O request, and load it into the new
+ *              selections in sorted order.
+ *
+ *              Note that in this case, it is the callers responsibility
```
```suggestion
 *              Note that in this case, it is the caller's responsibility
```

vchoi-hdfgroup committed:
Address spelling/grammar errors from PR review.
